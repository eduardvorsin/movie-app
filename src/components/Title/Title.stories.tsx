import type { Meta, StoryObj } from '@storybook/react';
import Title, { Props } from './Title';

const levels: Exclude<Props['level'], undefined>[] = [1, 2, 3, 4, 5, 6];
const weights: Exclude<Props['weight'], undefined>[] = [400, 500, 600, 700];

const meta: Meta<typeof Title> = {
	title: 'UI/Title',
	tags: ['autodocs'],
	argTypes: {
		className: {
			description: 'Additional classes for the Title'
		},
		testId: {
			description: 'Id for testing the component',
		},
		children: {
			description: 'Header Content',
		},
		level: {
			description: 'Header level',
		},
		as: {
			description: 'The HTML element that will be used on the page. A total of 6 elements are available from h1 to h6',
		}
	},
	component: Title,
	parameters: {
		docs: {
			description: {
				component: 'Utilize headings to designate pages or segments of a user interface. Headings have the flexibility to mark either a complete page or portions of associated content. With the Heading component, you can choose the suitable semantic element (h1-h6) and adjust its size independently to match the surrounding content appropriately.',
			},
		},
	},
}

export default meta;
type Story = StoryObj<typeof Title>;

export const Default: Story = {
	args: {
		level: 3,
		children: 'Default',
	},
	parameters: {
		docs: {
			description: {
				story: 'Default version of the header',
			},
		},
	},
};

export const Weights: Story = {
	parameters: {
		docs: {
			description: {
				story: 'There are 4 levels of weight(400-700) for this font',
			},
		},
	},
	render: () => (
		<div
			className='flex flex-col'
		>
			{weights.map((weight) => (
				<Title
					key={weight}
					className='mb-4 last:mb-4'
					weight={weight}
					level={3}
					as={`h${3}`}
				>
					Weight {weight}
				</Title>
			))}
		</div>
	)
};

export const Levels: Story = {
	parameters: {
		docs: {
			description: {
				story: 'There are 6 levels of headers, which differ in font size and line height',
			},
		},
	},
	render: () => (
		<div
			className='flex flex-col'
		>
			{levels.map((level) => (
				<Title
					key={level}
					className='mb-4 last:mb-4'
					level={level}
					as={`h${level}`}
				>
					Level {level}
				</Title>
			))}
		</div>
	)
};