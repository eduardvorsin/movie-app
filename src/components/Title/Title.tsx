import { GeneralProps, HeadingElement, HeadingLevel } from '@/types/shared';
import React from 'react';
import { twMerge } from '../../../tailwind.config';

export type Props = {
	children: React.ReactNode,
	level: HeadingLevel,
	as?: HeadingElement,
	weight?: 400 | 500 | 600 | 700,
	testId?: string,
} & GeneralProps;

const levelTypes = {
	1: 'text-400 sm:text-500 md:text-600 lg:text-700',
	2: 'text-400 sm:text-500 md:text-600',
	3: 'text-400 sm:text-500',
	4: 'text-300 sm:text-400',
	5: 'text-300',
	6: 'text-200',
} as const;

const fontWeights = {
	400: 'font-regular',
	500: 'font-medium',
	600: 'font-semibold',
	700: 'font-bold',
} as const;

export default function Title({
	level,
	children,
	className,
	testId,
	weight = 700,
	as = 'h6',
	...props
}: Props) {
	const classes = twMerge(
		'leading-[1.25] -tracking-[0.01em]',
		levelTypes[level],
		fontWeights[weight],
		className,
	);

	const Heading: HeadingElement = as;

	return (
		<Heading
			data-testid={testId}
			className={classes}
			{...props}
		>
			{children}
		</Heading>
	);
}