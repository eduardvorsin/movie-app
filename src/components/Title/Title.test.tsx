import { render, screen } from '@testing-library/react';
import Title from './Title';

describe('Title tests', () => {
	it('is rendered correctly', () => {
		render(<Title level={1}>Title</Title>)

		expect(screen.getByText<HTMLHeadingElement>(/title/i)).toBeInTheDocument();
	});

	it('If the prop Level is 4, then the text-400 class should be added', () => {
		render(<Title level={4}>Title</Title>)

		expect(screen.getByText<HTMLHeadingElement>(/title/i)).toHaveClass('text-400');
	});

	it('is a basic snapshot', () => {
		render(<Title level={4}>Title</Title>)

		expect(screen.getByText<HTMLHeadingElement>(/title/i)).toMatchSnapshot();
	});

	it('is a snapshot with header level 2', () => {
		render(<Title level={2}>Title</Title>)

		expect(screen.getByText<HTMLHeadingElement>(/title/i)).toMatchSnapshot();
	});

	it('is a snapshot with html element h5', () => {
		render(<Title as='h5' level={3}>Title</Title>)

		expect(screen.getByText<HTMLHeadingElement>(/title/i)).toMatchSnapshot();
	});
});