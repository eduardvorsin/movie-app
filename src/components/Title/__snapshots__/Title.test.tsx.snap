// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Title tests is a basic snapshot 1`] = `
<h6
  class="leading-[1.25] -tracking-[0.01em] text-300 sm:text-400 font-bold "
>
  Title
</h6>
`;

exports[`Title tests is a snapshot with header level 2 1`] = `
<h6
  class="leading-[1.25] -tracking-[0.01em] text-400 sm:text-500 md:text-600 font-bold "
>
  Title
</h6>
`;

exports[`Title tests is a snapshot with html element h5 1`] = `
<h5
  class="leading-[1.25] -tracking-[0.01em] text-400 sm:text-500 font-bold "
>
  Title
</h5>
`;
